File: Python3/1218. Longest Arithmetic Subsequence of Given Difference.py
class Solution:
    def longestSubsequence(self, arr: List[int], difference: int) -> int:
        n = len(arr)
        dp = collections.defaultdict(int)
        ans = 1

        for i, n in enumerate(arr):
            dp[n] = max(dp.get(n - difference, 0) + 1, dp[n])
            ans = max(ans, dp[n])

        return ans

File: Java/1218. Longest Arithmetic Subsequence of Given Difference.java
class Solution {
    public int longestSubsequence(int[] arr, int difference) {
        int ans = 1;
        Map<Integer, Integer> dp = new HashMap();
        
        for (int n: arr) {
            dp.put(n, Math.max(dp.getOrDefault(n, 0), 1 + dp.getOrDefault(n - difference, 0)));
            ans = Math.max(ans, dp.get(n));
        }

        return ans;
    }
}